src/workerd/util/perfetto-tracing.c++

```diff
@@ -0,0 +1,111 @@
+#include "perfetto-tracing.h"
+
+#if defined(WORKERD_USE_PERFETTO)
+
+#include <kj/debug.h>
+#include <kj/filesystem.h>
+#include <kj/memory.h>
+#include <kj/vector.h>
+#include <fcntl.h>
+#include <perfetto/tracing/track_event_legacy.h>
+#include "protos/perfetto/config/data_source_config.gen.h"
+#include "protos/perfetto/config/trace_config.gen.h"
+#include "protos/perfetto/config/track_event/track_event_config.gen.h"
+
+PERFETTO_TRACK_EVENT_STATIC_STORAGE_IN_NAMESPACE(workerd::traces);
+
+namespace workerd {
+namespace {
+kj::AutoCloseFd openTraceFile(kj::StringPtr path) {
+  int fd = open(path.cStr(), O_RDWR | O_CREAT | O_TRUNC, 0600);
+  KJ_REQUIRE(fd >= 0, "Unable to open tracing file");
+  return kj::AutoCloseFd(fd);
+}
+
+void initializePerfettoOnce() {
+  if (perfetto::Tracing::IsInitialized()) return;
+  perfetto::TracingInitArgs args;
+  args.backends |= perfetto::kInProcessBackend;
+  perfetto::Tracing::Initialize(args);
+  PerfettoSession::registerWorkerdTracks();
+}
+
+std::unique_ptr<perfetto::TracingSession> createTracingSession(int fd, kj::StringPtr categories) {
+  initializePerfettoOnce();
+  perfetto::protos::gen::TrackEventConfig track_event_cfg;
+  track_event_cfg.add_disabled_categories("*");
+
+  // The categories is a comma-separated list
+  auto cats = PerfettoSession::parseCategories(categories);
+  for (auto category : cats) {
+    auto view = std::string(category.begin(), category.size());
+    track_event_cfg.add_enabled_categories(view);
+  }
+
+  perfetto::TraceConfig cfg;
+  cfg.add_buffers()->set_size_kb(256);  // Record up to 1 MiB.
+  auto* ds_cfg = cfg.add_data_sources()->mutable_config();
+  ds_cfg->set_name("track_event");
+  ds_cfg->set_track_event_config_raw(track_event_cfg.SerializeAsString());
+  std::unique_ptr<perfetto::TracingSession> tracing_session(
+      perfetto::Tracing::NewTrace());
+  tracing_session->Setup(cfg, fd);
+  return kj::mv(tracing_session);
+}
+}  // namespace
+
+void PerfettoSession::registerWorkerdTracks() {
```

Comment:
If this is exposed for edgeworker's benefit, I ended up going with `workerd::traces::TrackEvent::Register()`.

Reply:
Ok, I'll leave it for now tho.